sale_import_base/components/importer_sale_channel.py

```diff
@@ -0,0 +1,337 @@
+
+    def run(self, raw_data):
+        """
+        :param raw_data: json-like string
+        :return: generated sale order
+        """
+        # we need this step because the chunk data is stored as
+        # a string (so we can edit it)
+        try:
+            so_datamodel_load = self.env.datamodels["sale.order"].load_json(raw_data)
+        except MarshmallowValidationError:
+            raise ValidationError(MarshmallowValidationError)
+        except Exception:
+            raise
+        json_data_initial = so_datamodel_load.dump()
+        json_data = deepcopy(json_data_initial)
+        so_vals = self.process_data(json_data)
+        new_sale_order = self.env["sale.order"].create(so_vals)
+        self.finalize(new_sale_order, json_data_initial)
+        return new_sale_order
+
+    # DATAMODEL PROCESSORS
+    def process_data(self, so_vals):
+        """ Transform values in-place
+         to make it usable in create() """
+        self.process_simple_fields(so_vals)
+        self.process_m2os(so_vals)
+        so_vals = self.simulate_onchanges(so_vals)
+        return so_vals
+
+    def process_simple_fields(self, so_vals):
+        del so_vals["currency_code"]
+        del so_vals["payment"]
+
+    def process_m2os(self, so_vals):
+        channel_id = self.get_channel(so_vals)
+        partner_id = self.get_partner(so_vals, channel_id)
+        self.process_addresses(partner_id, so_vals)
+        self.process_lines(so_vals)
+        self.process_amount(so_vals)
+        self.process_invoice(so_vals)
+        self.process_sale_channel(so_vals)
+
+    def get_channel(self, so_vals):
+        sale_channel = self.env["sale.channel"].search(
+            [("name", "=", so_vals["sale_channel"])]
+        )
+        return sale_channel
+
+    def get_partner(self, so_vals, sale_channel):
+        external_id = so_vals["address_customer"].get("external_id")
+        binding = self.env["sale.channel.partner"].search(
+            [
+                ("external_id", "=", external_id),
+                ("sale_channel_id", "=", sale_channel.id),
+            ]
+        )
+        if binding:
+            return binding.partner_id
+        if sale_channel.allow_match_on_email:
+            partner = self.env["res.partner"].search(
+                [("email", "=", so_vals["address_customer"]["email"])]
+            )
+            if partner:
+                return partner
+        return self.create_partner(so_vals)
+
+    def create_partner(self, so_vals):
+        partner_vals = dict()
+        for field in MAPPINGS_SALE_ORDER_ADDRESS_SIMPLE:
+            partner_vals[field] = so_vals["address_customer"][field]
+        country = self.env["res.country"].search(
+            [("code", "=", so_vals["address_customer"]["country_code"])]
+        )
+        partner_vals["country_id"] = country.id
+        partner = self.env["res.partner"].create(partner_vals)
+        return partner
+
+    def process_addresses(self, partner_id, so_vals):
```

Comment:
this method is complex.
I think we should not try to update the customer in the same method as the address, it make the code more complex.
Morevover in case of a new partner we create it and them we update it, a lot of sql request.